### PR TITLE
Plumb Okio stream types into request handlers.

### PR DESCRIPTION
**Pull request by JakeWharton**
This simplifies a ton of byte-management operations like checking for webp or even converting to a byte[].

@swankjesse

### CODE REVIEW

**Comment by JakeWharton** [comment]
Oh no it's that stupid Robolectric bug...

File: picasso/src/main/java/com/squareup/picasso/BitmapHunter.java [hunk 1]
```diff
@@ -119,37 +122,39 @@
...
       if (calculateSize) {
         BitmapFactory.decodeByteArray(bytes, 0, bytes.length, options);
         RequestHandler.calculateInSampleSize(request.targetWidth, request.targetHeight, options,
             request);
       }
       return BitmapFactory.decodeByteArray(bytes, 0, bytes.length, options);
     } else {
+      InputStream stream = bufferedSource.inputStream();
       if (calculateSize) {
+        // TODO use an InputStream that buffers with Okio...
```
Comment (swankjesse) [inline picasso/src/main/java/com/squareup/picasso/BitmapHunter.java]: yeah

File: picasso/src/main/java/com/squareup/picasso/Utils.java [hunk 2]
```diff
@@ -283,25 +276,15 @@ static boolean hasPermission(Context context, String permission) {
...
-    byte[] fileHeaderBytes = new byte[WEBP_FILE_HEADER_SIZE];
-    boolean isWebPFile = false;
-    if (stream.read(fileHeaderBytes, 0, WEBP_FILE_HEADER_SIZE) == WEBP_FILE_HEADER_SIZE) {
-      // If a file's header starts with RIFF and end with WEBP, the file is a WebP file
-      isWebPFile = WEBP_FILE_HEADER_RIFF.equals(new String(fileHeaderBytes, 0, 4, "US-ASCII"))
-          && WEBP_FILE_HEADER_WEBP.equals(new String(fileHeaderBytes, 8, 4, "US-ASCII"));
+  static boolean isWebPFile(BufferedSource source) throws IOException {
+    if (source.request(WEBP_FILE_HEADER_SIZE)) {
+      Buffer buffer = source.buffer();
+      if (buffer.rangeEquals(0, WEBP_FILE_HEADER_RIFF)
```
Comment (swankjesse) [inline picasso/src/main/java/com/squareup/picasso/Utils.java]: you could potentially call `source.rangeEquals()` right? Then you don’t even need to call `request()` above

Comment (swankjesse) [inline picasso/src/main/java/com/squareup/picasso/Utils.java]: ... and that won’t request a 2nd byte if the first byte isn’t `R` or W`

Comment (JakeWharton) [inline picasso/src/main/java/com/squareup/picasso/Utils.java]: Ah, yeah. I was so worried about not consuming the byes I forgot about that.

Comment (JakeWharton) [inline picasso/src/main/java/com/squareup/picasso/Utils.java]: https://github.com/square/picasso/pull/1593

**Pull request merged**